File: docs/harness-arguments.js
"use strict";

const HarnessArguments={
  values:new Map(),receipts:new Map(),program:null,
  edits(name){if(!this.values.has(name))this.values.set(name,new Map());return this.values.get(name);},
  inputs(name){
    const edits=this.edits(name),out=Object.create(null);let size=0;
    if(edits.size>64)throw Error("At most 64 invocation bindings");
    for(const [key,raw] of edits){
      if(!key||key.length>128)throw Error("Invalid argument name");
      size+=raw.length;if(size>65536)throw Error("Invocation bindings exceed 64 KiB");
      try{out[key]=JSON.parse(raw);}catch(_){throw Error("Invalid JSON for argument "+key);}
    }
    return out;
  },
  open(){
    this.program=Harness.selected;if(!this.program)return;
    this.render();document.getElementById("argumentInspector").showModal();
  },
  render(){
    const name=this.program,doc=Harness.document(name),edits=this.edits(name);
    document.getElementById("argumentTitle").textContent=name+" arguments";
    const definitions=new Map((doc.nodes||[]).filter(n=>n.type==="scope.in").map(n=>[n.params?.name?.replace(/\?$/, ""),n.params||{}]).filter(([key])=>key));
    for(const key of edits.keys())if(!definitions.has(key))definitions.set(key,{});
    const body=document.getElementById("argumentInputs");body.replaceChildren();
    for(const [key,param] of definitions){
      const row=document.createElement("tr"),toggle=document.createElement("input"),input=document.createElement("input");
      toggle.type="checkbox";toggle.checked=edits.has(key);toggle.setAttribute("aria-label","Supply "+key);
      input.setAttribute("aria-label","Value for "+key);input.spellcheck=false;
      input.value=edits.get(key)??JSON.stringify(param.default??"");input.disabled=!toggle.checked;
      toggle.addEventListener("change",()=>{input.disabled=!toggle.checked;if(toggle.checked)edits.set(key,input.value);else edits.delete(key);this.validate();});
      input.addEventListener("input",()=>{edits.set(key,input.value);this.validate();});
      for(const value of [toggle,key,param.kind||"generic",input,Object.hasOwn(param,"default")?JSON.stringify(param.default):"unbound"]){
        const cell=document.createElement("td");if(typeof value==="string")cell.textContent=value;else cell.append(value);row.append(cell);
      }
      body.append(row);
    }
    this.validate();this.renderReceipt();
  },
  validate(){
    let error="";try{this.inputs(this.program);}catch(e){error=e.message;}
    document.getElementById("argumentError").textContent=error;
    document.getElementById("argumentRun").disabled=!!error||Harness.running||Harness.dirty||Harness.selected!==this.program||Harness.inspectionOnly();
    return !error;
  },
  add(){
    const field=document.getElementById("argumentName"),name=field.value.trim();
    if(!name||name.length>128||this.edits(this.program).has(name))return;
    this.edits(this.program).set(name,'""');field.value="";this.render();
  },
  record(receipt){
    const entry=Harness.board[receipt?.programKey];
    if(!receipt?.program||!receipt.programCid||receipt.programCid!==entry?.programCid)return;
    const old=this.receipts.get(receipt.program);
    if(old){
      if(old.runId===receipt.runId&&Number(old.timelineRevision)>Number(receipt.timelineRevision))return;
      if(old.runId!==receipt.runId&&Number(old.startedAtMs)>Number(receipt.startedAtMs))return;
      if(old.startedAtMs===receipt.startedAtMs&&Number(old.sequence)>Number(receipt.sequence))return;
    }
    this.receipts.set(receipt.program,receipt);
    if(this.program===receipt.program&&document.getElementById("argumentInspector").open)this.renderReceipt();
  },
  renderReceipt(){
    const body=document.getElementById("argumentResolved");body.replaceChildren();
    const receipt=this.receipts.get(this.program),entry=Harness.board["lcnc/program/"+this.program];
    const valid=!!receipt?.programCid&&receipt.programCid===entry?.programCid&&!Harness.drafts.has(this.program);
    const status=document.getElementById("argumentReceipt");
    status.textContent=!receipt?"No recorded invocation":!valid?"Receipt belongs to another program version":
      receipt.status+" / "+receipt.runId+(receipt.error?" / "+receipt.error:"")+(receipt.bindingsTruncated?" / binding report truncated":"");
    if(!valid)return;
    const rows=(receipt.bindings||[]).map(r=>({...r,node:r.nodeId}));
    const nodes=[];const walk=ns=>{for(const n of ns||[]){nodes.push(n);walk(n.children);}};walk(entry.document?.nodes);
    for(const n of nodes.filter(n=>n.type==="ccek.incarnate"))for(const r of receipt.outputs?.[n.id]?.arguments||[])rows.push({...r,node:n.id});
    for(const r of rows.slice(0,1024)){
      const row=document.createElement("tr");
      for(const value of [r.node,r.name,r.type,r.source,JSON.stringify(r.value),r.status]){
        const cell=document.createElement("td");cell.textContent=value??"";
        if(r.overridden)cell.title="Overrides "+r.overridden.source+": "+JSON.stringify(r.overridden.value);
        row.append(cell);
      }body.append(row);
    }
  },
};

File: src/jsHarness/kotlin/borg/trikeshed/web/harness/HarnessArguments.kt
package borg.trikeshed.web.harness

import org.w3c.dom.HTMLButtonElement
import org.w3c.dom.HTMLElement
import org.w3c.dom.HTMLInputElement

/** harness-arguments.js: invocation bindings per program and the resolved-binding receipt. */
object HarnessArguments {
    private val values = HashMap<String, LinkedHashMap<String, String>>()
    private val receipts = HashMap<String, dynamic>()
    var program: String? = null

    fun edits(name: String): LinkedHashMap<String, String> = values.getOrPut(name) { LinkedHashMap() }

    fun inputs(name: String): dynamic {
        val edits = edits(name)
        val out: dynamic = js("Object.create(null)")
        var size = 0
        InvocationLimits.checkCount(edits.size)
        for ((key, raw) in edits) {
            InvocationLimits.checkName(key)
            size += raw.length; InvocationLimits.checkSize(size)
            try { out[key] = JSON.parse(raw) } catch (_: Throwable) { throw IllegalArgumentException("Invalid JSON for argument $key") }
        }
        return out
    }

    fun open() {
        program = Harness.selected; if (program == null) return
        render(); byId("argumentInspector").asDynamic().showModal()
    }

    fun render() {
        val name = program!!
        val doc = Harness.document(name); val edits = edits(name)
        byId("argumentTitle").textContent = "$name arguments"
        val definitions = LinkedHashMap<String, dynamic>()
        for (n in (if (truthy(doc.nodes)) doc.nodes.unsafeCast<Array<dynamic>>() else emptyArray())) {
            if (n.type != "scope.in") continue
            val raw: dynamic = n.params?.name
            val key = if (jsTypeOf(raw) == "string") InvocationLimits.bindingName(raw as String) else null
            if (!key.isNullOrEmpty()) definitions[key] = if (truthy(n.params)) n.params else jsObject()
        }
        for (key in edits.keys) if (!definitions.containsKey(key)) definitions[key] = jsObject()
        val body = byId("argumentInputs"); body.replaceKids()
        for ((key, param) in definitions) {
            val row = el("tr"); val toggle = el("input").unsafeCast<HTMLInputElement>(); val input = el("input").unsafeCast<HTMLInputElement>()
            toggle.type = "checkbox"; toggle.checked = edits.containsKey(key); toggle.setAttribute("aria-label", "Supply $key")
            input.setAttribute("aria-label", "Value for $key"); input.spellcheck = false
            input.value = edits[key] ?: JSON.stringify(if (param.default == null) "" else param.default); input.disabled = !toggle.checked
            toggle.addEventListener("change", { input.disabled = !toggle.checked; if (toggle.checked) edits[key] = input.value else edits.remove(key); validate() })
            input.addEventListener("input", { edits[key] = input.value; validate() })
            val cells: List<Any> = listOf(toggle, key, if (truthy(param.kind)) jsString(param.kind) else "generic", input,
                if (hasOwn(param, "default")) JSON.stringify(param.default) else "unbound")
            for (value in cells) {
                val cell = el("td"); if (value is String) cell.textContent = value else cell.append(value.unsafeCast<HTMLElement>()); row.append(cell)
            }
            body.append(row)
        }
        validate(); renderReceipt()
    }

    fun validate(): Boolean {
        var error = ""
        try { inputs(program!!) } catch (e: Throwable) { error = errorMessage(e) }
        byId("argumentError").textContent = error
        byId("argumentRun").unsafeCast<HTMLButtonElement>().disabled =
            error.isNotEmpty() || Harness.running || Harness.dirty || Harness.selected != program || Harness.inspectionOnly()
        return error.isEmpty()
    }

    fun add() {
        val field = byId("argumentName").unsafeCast<HTMLInputElement>(); val name = field.value.trim()
        if (!InvocationLimits.acceptsNew(name, edits(program!!).containsKey(name))) return
        edits(program!!)[name] = "\"\""; field.value = ""; render()
    }

    private fun stamp(r: dynamic) = ReceiptOrder.Stamp(r.runId, jsNumber(r.timelineRevision), r.startedAtMs, jsNumber(r.startedAtMs), jsNumber(r.sequence))

    fun record(receipt: dynamic) {
        if (!truthy(receipt) || !truthy(receipt.program) || !truthy(receipt.programCid)) return
        val entry: dynamic = Harness.board[receipt.programKey]
        if (receipt.programCid != entry?.programCid) return
        val name = receipt.program as String
        val old = receipts[name]
        if (old != null && ReceiptOrder.keepsOld(stamp(old), stamp(receipt))) return
        receipts[name] = receipt
        if (program == name && byId("argumentInspector").asDynamic().open == true) renderReceipt()
    }

    fun renderReceipt() {
        val body = byId("argumentResolved"); body.replaceKids()
        val name = program ?: ""
        val receipt = receipts[name]; val entry = Harness.board[HarnessKeys.PROGRAM + name]
        val valid = truthy(receipt?.programCid) && receipt.programCid == entry?.programCid && !Harness.drafts.has(name)
        val status = byId("argumentReceipt")
        status.textContent = if (receipt == null) "No recorded invocation" else if (!valid) "Receipt belongs to another program version" else
            jsString(receipt.status) + " / " + jsString(receipt.runId) + (if (truthy(receipt.error)) " / " + jsString(receipt.error) else "") + (if (truthy(receipt.bindingsTruncated)) " / binding report truncated" else "")
        if (!valid) return
        val rows = ArrayList<dynamic>()
        for (r in (if (truthy(receipt.bindings)) receipt.bindings.unsafeCast<Array<dynamic>>() else emptyArray())) {
            val copy: dynamic = js("Object.assign({}, r)"); copy.node = r.nodeId; rows.add(copy)
        }
        val nodes = ArrayList<dynamic>()
        fun walk(ns: dynamic) { if (!truthy(ns)) return; for (n in ns.unsafeCast<Array<dynamic>>()) { nodes.add(n); walk(n.children) } }
        walk(entry.document?.nodes)
        for (n in nodes.filter { it.type == "ccek.incarnate" }) {
            val outputs: dynamic = receipt.outputs
            val args: dynamic = if (outputs == null) null else outputs[n.id]?.arguments
            for (r in (if (truthy(args)) args.unsafeCast<Array<dynamic>>() else emptyArray())) { val copy: dynamic = js("Object.assign({}, r)"); copy.node = n.id; rows.add(copy) }
        }
        for (r in rows.take(InvocationLimits.RESOLVED_ROWS)) {
            val row = el("tr")
            for (value in listOf<dynamic>(r.node, r.name, r.type, r.source, JSON.stringify(r.value), r.status)) {
                val cell = el("td"); cell.textContent = if (value == null) "" else jsString(value)
                if (truthy(r.overridden)) cell.title = "Overrides " + jsString(r.overridden.source) + ": " + JSON.stringify(r.overridden.value)
                row.append(cell)
            }
            body.append(row)
        }
    }
}
